### DISCUSSION TITLE
[SOLVED] Possible to have multiple VENDOR interfaces?

### DISCUSSION THREAD

**Discussion by kripton** [hathach/tinyusb#664] [Q&A] [answered]
Hi, I'm trying to use the RP2040 (Pico Board) as a USB device using TinyUSB. When I have one CDC interface (= CMD interface + DATA interface) + 1 VENDOR interface, all is working well and as expected.
However, when I add a second VENDOR interface (in the same manner as the first, of course with different endpoints), the USB device fails to enumerate (on Linux) with this error in the kernel log: `usb 1-1: can't set config #1, error -32`. Is this expected or what would I need to to do have two (or more ;)) VENDOR interfaces?

Relevant code parts:
```
#ifndef CFG_TUD_ENDPOINT0_SIZE
#define CFG_TUD_ENDPOINT0_SIZE    64
#endif

//------------- CLASS -------------//
#define CFG_TUD_CDC             1
#define CFG_TUD_HID             0
#define CFG_TUD_MIDI            0
#define CFG_TUD_MSC             0
#define CFG_TUD_VENDOR          1

#define CFG_TUD_CDC_RX_BUFSIZE 256
#define CFG_TUD_CDC_TX_BUFSIZE 256

#define CFG_TUD_VENDOR_RX_BUFSIZE 640
#define CFG_TUD_VENDOR_TX_BUFSIZE 640
```
```
enum {
    ITF_NUM_CDC_CMD,
    ITF_NUM_CDC_DATA,
    ITF_NUM_PORT00,
    ITF_NUM_PORT01,
    ITF_NUM_TOTAL
};

#define  CONFIG_TOTAL_LEN  (TUD_CONFIG_DESC_LEN + TUD_CDC_DESC_LEN + TUD_VENDOR_DESC_LEN + TUD_VENDOR_DESC_LEN)

// Endpoints 0x00 and 0x80 are for CONTROL transfers => do not use

#define USBD_CDC_EP_CMD          0x81
#define USBD_CDC_EP_OUT          0x02
#define USBD_CDC_EP_IN           0x82
#define USBD_CDC_CMD_MAX_SIZE       8
#define USBD_CDC_IN_OUT_MAX_SIZE   64

#define USBD_PORT00_EP_OUT       0x03
#define USBD_PORT00_EP_IN        0x83
#define USBD_PORT01_EP_OUT       0x04
#define USBD_PORT01_EP_IN        0x84


uint8_t const desc_configuration[] =
{
    // Config number, interface count, string index, total length, attribute, power in mA
    TUD_CONFIG_DESCRIPTOR(1, ITF_NUM_TOTAL, 0, CONFIG_TOTAL_LEN,
        TUSB_DESC_CONFIG_ATT_REMOTE_WAKEUP, 500),

    // Interface number, string index, EP notification address and size, EP data address (out, in) and size.
    TUD_CDC_DESCRIPTOR(ITF_NUM_CDC_CMD, 4, USBD_CDC_EP_CMD,
        USBD_CDC_CMD_MAX_SIZE, USBD_CDC_EP_OUT, USBD_CDC_EP_IN,
        USBD_CDC_IN_OUT_MAX_SIZE),

    // Interface number, string index, EP Out & IN address, EP size
    TUD_VENDOR_DESCRIPTOR(ITF_NUM_PORT00, 5, USBD_PORT00_EP_OUT, USBD_PORT00_EP_IN, 64),
    TUD_VENDOR_DESCRIPTOR(ITF_NUM_PORT01, 6, USBD_PORT01_EP_OUT, USBD_PORT01_EP_IN, 64),
};
```

**Comment by kripton** [comment, marked as answer]
Okay, that was quite easy. I looked at some more examples and the solution is:
`#define CFG_TUD_VENDOR          2`
I thought those were feature-flags if I want to have support for VENDOR interfaces or not. Didn't see that this is the number of interfaces per class.

Thanks for this awesome library!

**Reply by hathach** [reply]
I really need to write more document 😁😁

**Reply by pkoning2** [reply]
I found there's a second part of the answer: while it isn't obvious this is required, the endpoint numbers need to be different for the two interfaces.

**Comment by JakeSays** [comment]
Just hit this same issue, and after three hours I find this discussion.

Suggestion: put the documentation in the code:
```#define CFG_TUD_HID_INSTANCES = 1```

**Reply by hathach** [reply]
good proposal but it is too late now. And people get familiar with it anyway. We only need to add a couple of comments in example configure file.

**Reply by JakeSays** [reply]
Which example config file? Also, the config file I was looking at was from adafruit's tinyusb wrapper (which, btw, is harder to use than your C api). Problem is you can't fix all the config files out there. Perhaps the documentation should go in tusb_option.h.

Anyway, you're right - people do get used to it, and now that I understand it tusb is working excellently:) Thanks for a great library.
